File: figcli/fig-ipc/command.go
package fig_ipc

import (
	fig_proto "fig-cli/fig-proto"
	"fmt"
	"os"
	"time"

	"google.golang.org/protobuf/proto"
)

func SendCommand(cmd *fig_proto.Command) error {
	conn, err := Connect()
	if err != nil {
		return err
	}
	defer conn.Close()

	message := fig_proto.LocalMessage{
		Type: &fig_proto.LocalMessage_Command{
			Command: cmd,
		},
	}

	if err := conn.SendFigProto(&message); err != nil {
		return err
	}

	return nil
}

func SendRecvCommand(cmd *fig_proto.Command) (*fig_proto.CommandResponse, error) {
	conn, err := Connect()
	if err != nil {
		return nil, err
	}
	defer conn.Close()

	message := fig_proto.LocalMessage{
		Type: &fig_proto.LocalMessage_Command{
			Command: cmd,
		},
	}

	if err = conn.SendFigProto(&message); err != nil {
		return nil, err
	}

	msg := conn.RecvMessageTimeout(time.Second * 3)
	if msg.Error != nil {
		return nil, msg.Error
	}

	if msg.ProtoType != protoTypeFigProto {
		return nil, fmt.Errorf("unexpected message type: %d", msg.ProtoType)
	}

	var cmdResponse fig_proto.CommandResponse
	proto.Unmarshal(msg.Message, &cmdResponse)
	return &cmdResponse, nil
}

func GetCommandResponseMessage(commandResponse *fig_proto.CommandResponse) (string, error) {
	switch commandResponse.Response.(type) {
	case *fig_proto.CommandResponse_Success:
		return commandResponse.GetSuccess().GetMessage(), nil
	case *fig_proto.CommandResponse_Error:
		return commandResponse.GetError().GetMessage(), nil
	default:
		return "", fmt.Errorf("unknown response %T", commandResponse.Response)
	}
}

func RestartCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_Restart{
			Restart: &fig_proto.RestartCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func QuitCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_Quit{
			Quit: &fig_proto.QuitCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func UpdateCommand(force bool) error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_Update{
			Update: &fig_proto.UpdateCommand{
				Force: force,
			},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func ReportWindowCommand(message string) error {
	path := os.Getenv("PATH")
	figEnvVar := os.Getenv("FIG_ENV_VAR")
	term := os.Getenv("TERM")

	cmd := fig_proto.Command{
		Command: &fig_proto.Command_ReportWindow{
			ReportWindow: &fig_proto.ReportWindowCommand{
				Report:    message,
				Path:      path,
				FigEnvVar: figEnvVar,
				Terminal:  term,
			},
		},
	}

	err := SendCommand(&cmd)
	if err != nil {
		return err
	}

	return nil
}

func RestartSettingsListenerCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_RestartSettingsListener{
			RestartSettingsListener: &fig_proto.RestartSettingsListenerCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func RunInstallScriptCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_RunInstallScript{
			RunInstallScript: &fig_proto.RunInstallScriptCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func RunBuildCommand(branch string) error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_Build{
			Build: &fig_proto.BuildCommand{
				Branch: &branch,
			},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func RunOpenUiElementCommand(element fig_proto.UiElement) (string, error) {
	cmd := fig_proto.Command{
		Command: &fig_proto.Command_OpenUiElement{
			OpenUiElement: &fig_proto.OpenUiElementCommand{
				Element: element,
			},
		},
	}

	response, err := SendRecvCommand(&cmd)
	if err != nil {
		return "", err
	}

	return GetCommandResponseMessage(response)
}

func RunResetCacheCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_ResetCache{
			ResetCache: &fig_proto.ResetCacheCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func ToggleDebugModeCommand() (string, error) {
	toggle := true

	cmd := fig_proto.Command{
		Command: &fig_proto.Command_DebugMode{
			DebugMode: &fig_proto.DebugModeCommand{
				ToggleDebugMode: &toggle,
			},
		},
	}

	response, err := SendRecvCommand(&cmd)
	if err != nil {
		return "", err
	}

	return GetCommandResponseMessage(response)
}

func SetDebugModeCommand(debugMode bool) (string, error) {
	cmd := fig_proto.Command{
		Command: &fig_proto.Command_DebugMode{
			DebugMode: &fig_proto.DebugModeCommand{
				SetDebugMode: &debugMode,
			},
		},
	}

	response, err := SendRecvCommand(&cmd)
	if err != nil {
		return "", err
	}

	return GetCommandResponseMessage(response)
}

func GetDebugModeCommand() (string, error) {
	cmd := fig_proto.Command{
		Command: &fig_proto.Command_DebugMode{
			DebugMode: &fig_proto.DebugModeCommand{},
		},
	}

	response, err := SendRecvCommand(&cmd)
	if err != nil {
		return "", err
	}

	return GetCommandResponseMessage(response)
}

func PromptAccessibilityCommand() error {
	noResponse := true

	cmd := fig_proto.Command{
		NoResponse: &noResponse,
		Command: &fig_proto.Command_PromptAccessibility{
			PromptAccessibility: &fig_proto.PromptAccessibilityCommand{},
		},
	}

	if err := SendCommand(&cmd); err != nil {
		return err
	}

	return nil
}

func CreateInputMethodRequest(
	action fig_proto.InputMethodAction,
) *fig_proto.Command {
	id := int64(0)
	noResponse := false

	return &fig_proto.Command{
		Id:         &id,
		NoResponse: &noResponse,
		Command: &fig_proto.Command_InputMethod{
			InputMethod: &fig_proto.InputMethodCommand{
				Actions: &action,
			},
		},
	}
}

func InputMethodCommand(command string) (string, error) {
	cmd := CreateInputMethodRequest(fig_proto.InputMethodAction_STATUS_OF_INPUT_METHOD)

	switch command {
	case "install":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_INSTALL_INPUT_METHOD)
	case "uninstall":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_UNINSTALL_INPUT_METHOD)
	case "enable":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_ENABLE_INPUT_METHOD)
	case "disable":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_DISABLE_INPUT_METHOD)
	case "select":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_SELECT_INPUT_METHOD)
	case "deselect":
		cmd = CreateInputMethodRequest(fig_proto.InputMethodAction_DESELECT_INPUT_METHOD)
	default:
		break
	}

	res, err := SendRecvCommand(cmd)
	if err != nil {
		return "", err
	}

	return GetCommandResponseMessage(res)

}

File: dotfilesd/src/ipc/command.rs
use super::{send_command_to_socket, send_recv_command_to_socket};
use anyhow::Result;
use fig_proto::local;
use fig_proto::local::{
    command, BuildCommand, DebugModeCommand, InputMethodAction, InputMethodCommand,
    OpenUiElementCommand, PromptAccessibilityCommand, QuitCommand, RestartCommand,
    RestartSettingsListenerCommand, UiElement, UpdateCommand,
};

pub async fn restart_settings_listener() -> Result<()> {
    let command = command::Command::RestartSettingsListener(RestartSettingsListenerCommand {});
    send_command_to_socket(command).await
}

pub async fn open_ui_element(element: UiElement) -> Result<()> {
    let command = command::Command::OpenUiElement(OpenUiElementCommand {
        element: element.into(),
    });
    send_command_to_socket(command).await
}

pub async fn run_build_command(build: impl Into<String>) -> Result<()> {
    let command = command::Command::Build(BuildCommand {
        branch: Some(build.into()),
    });
    send_command_to_socket(command).await
}

pub async fn toggle_debug_mode() -> Result<local::CommandResponse> {
    let command = command::Command::DebugMode(DebugModeCommand {
        set_debug_mode: None,
        toggle_debug_mode: Some(true),
    });
    send_recv_command_to_socket(command).await
}

pub async fn set_debug_mode(debug_mode: bool) -> Result<local::CommandResponse> {
    let command = command::Command::DebugMode(DebugModeCommand {
        set_debug_mode: Some(debug_mode),
        toggle_debug_mode: None,
    });
    send_recv_command_to_socket(command).await
}

pub async fn input_method_command(action: InputMethodAction) -> Result<()> {
    let command = command::Command::InputMethod(InputMethodCommand {
        actions: Some(action.into()),
    });
    send_command_to_socket(command).await
}

pub async fn prompt_accessibility_command() -> Result<()> {
    let command = command::Command::PromptAccessibility(PromptAccessibilityCommand {});
    send_command_to_socket(command).await
}

pub async fn update_command(force: bool) -> Result<()> {
    let command = command::Command::Update(UpdateCommand { force });
    send_command_to_socket(command).await
}

pub async fn restart_command() -> Result<()> {
    let command = command::Command::Restart(RestartCommand {});
    send_command_to_socket(command).await
}

pub async fn quit_command() -> Result<()> {
    let command = command::Command::Quit(QuitCommand {});
    send_command_to_socket(command).await
}

/*
func ReportWindowCommand(message string) error {
    path := os.Getenv("PATH")
    figEnvVar := os.Getenv("FIG_ENV_VAR")
    term := os.Getenv("TERM")

    cmd := fig_proto.Command{
        Command: &fig_proto.Command_ReportWindow{
            ReportWindow: &fig_proto.ReportWindowCommand{
                Report:    message,
                Path:      path,
                FigEnvVar: figEnvVar,
                Terminal:  term,
            },
        },
    }

    err := SendCommand(&cmd)
    if err != nil {
        return err
    }

    return nil
}

func RunInstallScriptCommand() error {
    noResponse := true

    cmd := fig_proto.Command{
        NoResponse: &noResponse,
        Command: &fig_proto.Command_RunInstallScript{
            RunInstallScript: &fig_proto.RunInstallScriptCommand{},
        },
    }

    if err := SendCommand(&cmd); err != nil {
        return err
    }

    return nil
}

func RunResetCacheCommand() error {
    noResponse := true

    cmd := fig_proto.Command{
        NoResponse: &noResponse,
        Command: &fig_proto.Command_ResetCache{
            ResetCache: &fig_proto.ResetCacheCommand{},
        },
    }

    if err := SendCommand(&cmd); err != nil {
        return err
    }

    return nil
}

func GetDebugModeCommand() (string, error) {
    cmd := fig_proto.Command{
        Command: &fig_proto.Command_DebugMode{
            DebugMode: &fig_proto.DebugModeCommand{},
        },
    }

    response, err := SendRecvCommand(&cmd)
    if err != nil {
        return "", err
    }

    return GetCommandResponseMessage(response)
}

*/
